// ngx-country-select/dist/index.js
import { Directive, ElementRef, NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import 'country-select-js';

var NgxCountrySelectDirective = (function () {
    /**
     * @param {?} el
     */
    function NgxCountrySelectDirective(el) {
        this.el = el;
        $ = $ || jQuery;
    }
    /**
     * @return {?}
     */
    NgxCountrySelectDirective.prototype.ngAfterViewInit = function () {
        this.$countrySelect = $(this.el.nativeElement).countrySelect();
    };
    /**
     * @param {?} countryCode
     * @return {?}
     */
    NgxCountrySelectDirective.prototype.setCountry = function (countryCode) {
        this.$countrySelect.countrySelect('selectCountry', countryCode.toLowerCase());
    };
    /**
     * @return {?}
     */
    NgxCountrySelectDirective.prototype.getCountryData = function () {
        return this.$countrySelect.countrySelect('getSelectedCountryData');
    };
    return NgxCountrySelectDirective;
}());
NgxCountrySelectDirective.decorators = [
    { type: Directive, args: [{
                selector: '[ngxCountrySelect]',
                exportAs: 'ngx-country-select'
            },] },
];
/**
 * @nocollapse
 */
NgxCountrySelectDirective.ctorParameters = function () { return [
    { type: ElementRef, },
]; };

var NgxCountrySelectModule = (function () {
    function NgxCountrySelectModule() {
    }
    return NgxCountrySelectModule;
}());
NgxCountrySelectModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    NgxCountrySelectDirective
                ],
                exports: [
                    NgxCountrySelectDirective
                ]
            },] },
];
/**
 * @nocollapse
 */
NgxCountrySelectModule.ctorParameters = function () { return []; };

export { NgxCountrySelectModule, NgxCountrySelectDirective };
