// ngx-country-select/dist/ngx-country-select.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('country-select-js')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/common', 'country-select-js'], factory) :
	(factory((global['ngx-country-select'] = {}),global._angular_core,global._angular_common));
}(this, (function (exports,_angular_core,_angular_common) { 'use strict';

var NgxCountrySelectDirective = (function () {
    /**
     * @param {?} el
     */
    function NgxCountrySelectDirective(el) {
        this.el = el;
        $ = $ || jQuery;
    }
    /**
     * @return {?}
     */
    NgxCountrySelectDirective.prototype.ngAfterViewInit = function () {
        this.$countrySelect = $(this.el.nativeElement).countrySelect();
    };
    /**
     * @param {?} countryCode
     * @return {?}
     */
    NgxCountrySelectDirective.prototype.setCountry = function (countryCode) {
        this.$countrySelect.countrySelect('selectCountry', countryCode.toLowerCase());
    };
    /**
     * @return {?}
     */
    NgxCountrySelectDirective.prototype.getCountryData = function () {
        return this.$countrySelect.countrySelect('getSelectedCountryData');
    };
    return NgxCountrySelectDirective;
}());
NgxCountrySelectDirective.decorators = [
    { type: _angular_core.Directive, args: [{
                selector: '[ngxCountrySelect]',
                exportAs: 'ngx-country-select'
            },] },
];
/**
 * @nocollapse
 */
NgxCountrySelectDirective.ctorParameters = function () { return [
    { type: _angular_core.ElementRef, },
]; };

var NgxCountrySelectModule = (function () {
    function NgxCountrySelectModule() {
    }
    return NgxCountrySelectModule;
}());
NgxCountrySelectModule.decorators = [
    { type: _angular_core.NgModule, args: [{
                imports: [
                    _angular_common.CommonModule
                ],
                declarations: [
                    NgxCountrySelectDirective
                ],
                exports: [
                    NgxCountrySelectDirective
                ]
            },] },
];
/**
 * @nocollapse
 */
NgxCountrySelectModule.ctorParameters = function () { return []; };

exports.NgxCountrySelectModule = NgxCountrySelectModule;
exports.NgxCountrySelectDirective = NgxCountrySelectDirective;

Object.defineProperty(exports, '__esModule', { value: true });

})));
