// ngx-country-picker/esm5/country-picker.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, ChangeDetectorRef, Component, Input, OnInit } from '@angular/core';
import { CountryPickerService } from './country-picker.service';
var CountryPickerComponent = /** @class */ (function () {
    function CountryPickerComponent(_countryPickerService, _cdr) {
        this._countryPickerService = _countryPickerService;
        this._cdr = _cdr;
        this.flag = false;
        this.setValue = 'cca3';
        this.setName = 'name.common';
        this.classes = ['form-control', 'form-control-sm'];
        this.countries = [];
        this.baseUrl = _countryPickerService.getBaseUrl() + 'data/';
    }
    CountryPickerComponent.prototype.ngOnInit = function () {
        var _this = this;
        this._countryPickerService.getCountries()
            .subscribe(function (countries) {
            _this.countries = countries.sort(function (a, b) {
                var na = _this.getName(a);
                var nb = _this.getName(b);
                if (na > nb) {
                    return 1;
                }
                if (na < nb) {
                    return -1;
                }
                return 0;
            });
            _this._cdr.markForCheck();
        });
    };
    CountryPickerComponent.prototype.getValue = function (obj) {
        return this.setValue.split('.').reduce(function (o, i) { return o[i]; }, obj);
    };
    CountryPickerComponent.prototype.getName = function (obj) {
        return this.setName.split('.').reduce(function (o, i) { return o[i]; }, obj);
    };
    CountryPickerComponent.ctorParameters = function () { return [
        { type: CountryPickerService },
        { type: ChangeDetectorRef }
    ]; };
    tslib_1.__decorate([
        Input()
    ], CountryPickerComponent.prototype, "flag", void 0);
    tslib_1.__decorate([
        Input()
    ], CountryPickerComponent.prototype, "setValue", void 0);
    tslib_1.__decorate([
        Input()
    ], CountryPickerComponent.prototype, "setName", void 0);
    tslib_1.__decorate([
        Input()
    ], CountryPickerComponent.prototype, "classes", void 0);
    CountryPickerComponent = tslib_1.__decorate([
        Component({
            selector: 'country-picker',
            template: "\n      <select [class]=\"classes\">\n          <option *ngFor=\"let c of countries\" [value]=\"getValue(c)\">\n              <img *ngIf=\"flag\" [src]=\"baseUrl + c.cca3.toLowerCase() + '.svg'\">{{ getName(c) }}\n          </option>\n      </select>\n  ",
            changeDetection: ChangeDetectionStrategy.OnPush
        })
    ], CountryPickerComponent);
    return CountryPickerComponent;
}());
export { CountryPickerComponent };
