// ngx-country-picker/__ivy_ngcc__/fesm2015/ngx-country-picker.js
import { __decorate, __param } from 'tslib';
import { InjectionToken, Inject, Injectable, ChangeDetectorRef, Input, Component, ChangeDetectionStrategy, Optional, SkipSelf, NgModule } from '@angular/core';
import { HttpResponse, HttpClient, HttpClientModule } from '@angular/common/http';
import { CommonModule } from '@angular/common';
import { throwError } from 'rxjs';
import { map, catchError } from 'rxjs/operators';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common/http';
import * as ɵngcc2 from '@angular/common';

function CountryPickerComponent_option_1_img_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "img", 3);
} if (rf & 2) {
    const c_r1 = ɵngcc0.ɵɵnextContext().$implicit;
    const ctx_r2 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("src", ctx_r2.baseUrl + c_r1.cca3.toLowerCase() + ".svg", ɵngcc0.ɵɵsanitizeUrl);
} }
function CountryPickerComponent_option_1_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "option", 1);
    ɵngcc0.ɵɵtemplate(1, CountryPickerComponent_option_1_img_1_Template, 1, 1, "img", 2);
    ɵngcc0.ɵɵtext(2);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const c_r1 = ctx.$implicit;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("value", ctx_r0.getValue(c_r1));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngIf", ctx_r0.flag);
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵtextInterpolate1("", ctx_r0.getName(c_r1), " ");
} }
const COUNTRY_PICKER_CONFIG = new InjectionToken('countrypicker.config');
const COUNTRY_PICKER_CONFIG_DEFAULT = {
    baseUrl: 'assets/',
    filename: 'countries.json',
};

var CountryPickerService_1;
let CountryPickerService = CountryPickerService_1 = class CountryPickerService {
    constructor(config, _http) {
        this._http = _http;
        this._data = null;
        this._baseUrl = config.baseUrl;
        this._filename = config.filename;
        this._data = this._loadData();
    }
    static handleError(error) {
        let errMsg;
        if (error instanceof HttpResponse) {
            if (error.status === 404) {
                errMsg = 'Error loading countries.json file.'
                    + ' Please configure WebPack and load countries.json assets to your root folder';
            }
            else {
                const body = error.body || '';
                const err = body.error || JSON.stringify(body);
                errMsg = `${error.status} - ${error.statusText || ''} ${err}`;
            }
        }
        else {
            errMsg = error.message ? error.message : error.toString();
        }
        console.error(errMsg);
        return throwError(errMsg);
    }
    getCountries() {
        return this._data
            .pipe(map((countries) => countries.map((country) => {
            country.name.native[0] = country.name.native[Object.keys(country.name.native)[0]];
            return country;
        })));
    }
    getBaseUrl() {
        return this._baseUrl;
    }
    _loadData() {
        return this._http.get(this._baseUrl + this._filename)
            .pipe(catchError(CountryPickerService_1.handleError));
    }
};
CountryPickerService.ɵfac = function CountryPickerService_Factory(t) { return new (t || CountryPickerService)(ɵngcc0.ɵɵinject(COUNTRY_PICKER_CONFIG), ɵngcc0.ɵɵinject(ɵngcc1.HttpClient)); };
CountryPickerService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: CountryPickerService, factory: CountryPickerService.ɵfac });
CountryPickerService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [COUNTRY_PICKER_CONFIG,] }] },
    { type: HttpClient }
];
CountryPickerService = CountryPickerService_1 = __decorate([ __param(0, Inject(COUNTRY_PICKER_CONFIG))
], CountryPickerService);

let CountryPickerComponent = class CountryPickerComponent {
    constructor(_countryPickerService, _cdr) {
        this._countryPickerService = _countryPickerService;
        this._cdr = _cdr;
        this.flag = false;
        this.setValue = 'cca3';
        this.setName = 'name.common';
        this.classes = ['form-control', 'form-control-sm'];
        this.countries = [];
        this.baseUrl = _countryPickerService.getBaseUrl() + 'data/';
    }
    ngOnInit() {
        this._countryPickerService.getCountries()
            .subscribe(countries => {
            this.countries = countries.sort((a, b) => {
                let na = this.getName(a);
                let nb = this.getName(b);
                if (na > nb) {
                    return 1;
                }
                if (na < nb) {
                    return -1;
                }
                return 0;
            });
            this._cdr.markForCheck();
        });
    }
    getValue(obj) {
        return this.setValue.split('.').reduce((o, i) => o[i], obj);
    }
    getName(obj) {
        return this.setName.split('.').reduce((o, i) => o[i], obj);
    }
};
CountryPickerComponent.ɵfac = function CountryPickerComponent_Factory(t) { return new (t || CountryPickerComponent)(ɵngcc0.ɵɵdirectiveInject(CountryPickerService), ɵngcc0.ɵɵdirectiveInject(ɵngcc0.ChangeDetectorRef)); };
CountryPickerComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: CountryPickerComponent, selectors: [["country-picker"]], inputs: { flag: "flag", setValue: "setValue", setName: "setName", classes: "classes" }, decls: 2, vars: 3, consts: [[3, "value", 4, "ngFor", "ngForOf"], [3, "value"], [3, "src", 4, "ngIf"], [3, "src"]], template: function CountryPickerComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "select");
        ɵngcc0.ɵɵtemplate(1, CountryPickerComponent_option_1_Template, 3, 3, "option", 0);
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵclassMap(ctx.classes);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.countries);
    } }, directives: [ɵngcc2.NgForOf, ɵngcc2.NgIf], encapsulation: 2, changeDetection: 0 });
CountryPickerComponent.ctorParameters = () => [
    { type: CountryPickerService },
    { type: ChangeDetectorRef }
];
__decorate([
    Input()
], CountryPickerComponent.prototype, "flag", void 0);
__decorate([
    Input()
], CountryPickerComponent.prototype, "setValue", void 0);
__decorate([
    Input()
], CountryPickerComponent.prototype, "setName", void 0);
__decorate([
    Input()
], CountryPickerComponent.prototype, "classes", void 0);

var CountryPickerModule_1;
function countryPickerServiceFactory(config, http) {
    return new CountryPickerService(config, http);
}
let CountryPickerModule = CountryPickerModule_1 = class CountryPickerModule {
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('CountryPickerModule is already loaded. Import it in the AppModule only');
        }
    }
    static forRoot(providedConfig = COUNTRY_PICKER_CONFIG_DEFAULT) {
        return {
            ngModule: CountryPickerModule_1,
            providers: [
                { provide: COUNTRY_PICKER_CONFIG, useValue: providedConfig },
                {
                    provide: CountryPickerService,
                    useFactory: countryPickerServiceFactory,
                    deps: [COUNTRY_PICKER_CONFIG, HttpClient]
                }
            ]
        };
    }
};
CountryPickerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CountryPickerModule });
CountryPickerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CountryPickerModule_Factory(t) { return new (t || CountryPickerModule)(ɵngcc0.ɵɵinject(CountryPickerModule, 12)); }, imports: [[
            CommonModule,
            HttpClientModule
        ]] });
CountryPickerModule.ctorParameters = () => [
    { type: CountryPickerModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
CountryPickerModule = CountryPickerModule_1 = __decorate([ __param(0, Optional()), __param(0, SkipSelf())
], CountryPickerModule);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountryPickerService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [COUNTRY_PICKER_CONFIG]
            }] }, { type: ɵngcc1.HttpClient }]; }, null); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountryPickerComponent, [{
        type: Component,
        args: [{
                selector: 'country-picker',
                template: `
      <select [class]="classes">
          <option *ngFor="let c of countries" [value]="getValue(c)">
              <img *ngIf="flag" [src]="baseUrl + c.cca3.toLowerCase() + '.svg'">{{ getName(c) }}
          </option>
      </select>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush
            }]
    }], function () { return [{ type: CountryPickerService }, { type: ɵngcc0.ChangeDetectorRef }]; }, { flag: [{
            type: Input
        }], setValue: [{
            type: Input
        }], setName: [{
            type: Input
        }], classes: [{
            type: Input
        }] }); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CountryPickerModule, { declarations: function () { return [CountryPickerComponent]; }, imports: function () { return [CommonModule,
        HttpClientModule]; }, exports: function () { return [CountryPickerComponent]; } }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CountryPickerModule, [{
        type: NgModule,
        args: [{
                imports: [
                    CommonModule,
                    HttpClientModule
                ],
                declarations: [
                    CountryPickerComponent
                ],
                exports: [
                    CountryPickerComponent
                ]
            }]
    }], function () { return [{ type: CountryPickerModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();

/**
 * Generated bundle index. Do not edit.
 */

export { COUNTRY_PICKER_CONFIG, COUNTRY_PICKER_CONFIG_DEFAULT, CountryPickerComponent, CountryPickerModule, CountryPickerService, countryPickerServiceFactory as ɵa };
