// ngx-country-picker/esm2015/country-picker.module.js
import * as tslib_1 from "tslib";
var CountryPickerModule_1;
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { HttpClient, HttpClientModule } from '@angular/common/http';
import { CommonModule } from '@angular/common';
import { COUNTRY_PICKER_CONFIG, COUNTRY_PICKER_CONFIG_DEFAULT } from './country-picker.config';
import { CountryPickerService } from './country-picker.service';
import { CountryPickerComponent } from './country-picker.component';
export function countryPickerServiceFactory(config, http) {
    return new CountryPickerService(config, http);
}
let CountryPickerModule = CountryPickerModule_1 = class CountryPickerModule {
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('CountryPickerModule is already loaded. Import it in the AppModule only');
        }
    }
    static forRoot(providedConfig = COUNTRY_PICKER_CONFIG_DEFAULT) {
        return {
            ngModule: CountryPickerModule_1,
            providers: [
                { provide: COUNTRY_PICKER_CONFIG, useValue: providedConfig },
                {
                    provide: CountryPickerService,
                    useFactory: countryPickerServiceFactory,
                    deps: [COUNTRY_PICKER_CONFIG, HttpClient]
                }
            ]
        };
    }
};
CountryPickerModule.ctorParameters = () => [
    { type: CountryPickerModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
CountryPickerModule = CountryPickerModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [
            CommonModule,
            HttpClientModule
        ],
        declarations: [
            CountryPickerComponent
        ],
        exports: [
            CountryPickerComponent
        ]
    }),
    tslib_1.__param(0, Optional()), tslib_1.__param(0, SkipSelf())
], CountryPickerModule);
export { CountryPickerModule };
